File: Java/61.java
__________________________________________________________________________________________________
sample 0 ms submission
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return null;
        ListNode start = head;
        int count = 1;
        while(start.next != null) {
            start = start.next;
            count++;
        }
        start.next = head;
        for(int i = count - k % count - 1; i > 0; i--) {
            head = head.next;
        }
        start = head.next;
        head.next = null;
        return start;
    }
}
__________________________________________________________________________________________________
sample 35348 kb submission
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head==null||k==0) return head;
        
        int len = 0;
        ListNode p = head;
        for(int i = 0;i < k; ++i){
            if(p!=null){
                ++len;
                p = p.next;
            }else{
                break;
            }
        }
        if(len==k&&p==null){
            return head;
        }
        
        
        if(len<k){
            k %= len;
            if(k==0){
                return head;
            }
            p = head;
            for(int i = 0;i < k; ++i){
                p = p.next;
            }
        }
        ListNode b = head;
        while(p.next!=null){
            p = p.next;
            b = b.next;
        }
        p.next = head;
        ListNode r = b.next;
        b.next = null;
        return r;
    }
}
__________________________________________________________________________________________________

File: Python3/61.py
__________________________________________________________________________________________________
sample 28 ms submission
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        last = head
        length = 1
        while last.next:
            last = last.next
            length = length + 1
        k = k % length
        last.next = head
        temp = head
        for _ in range(length - k - 1):
            temp = temp.next
        answer = temp.next
        temp.next = None
        
        return answer
__________________________________________________________________________________________________
sample 13028 kb submission
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if (head == None):
            return None
        
        list = [head]
        n = 1
        current_node = head
        result = head
        
        while (current_node.next != None):
            n = n + 1
            current_node = current_node.next
            list.append(current_node)
        
        current_node.next = head
        rotations = n - k % n
        
        if (rotations != 0):
            current_node = list[rotations - 1]
            result = current_node.next
            current_node.next = None
        
        return result
__________________________________________________________________________________________________
